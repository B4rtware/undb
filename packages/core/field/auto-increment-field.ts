import type { IAutoIncrementFilter } from '../filter/auto-increment.filter.js'
import type { IAutoIncrementFilterOperator } from '../filter/operators.js'
import { AutoIncrementFieldValue } from './auto-increment-field-value.js'
import type {
  AutoIncrementFieldType,
  ICreateAutoIncrementFieldInput,
  ICreateAutoIncrementFieldValue,
} from './auto-increment-field.type.js'
import { BaseField } from './field.base.js'
import type { IAutoIncrementField } from './field.type.js'
import type { IFieldVisitor } from './field.visitor.js'
import { FieldId, FieldName, FieldValueConstraints } from './value-objects/index.js'

export class AutoIncrementField extends BaseField<IAutoIncrementField> {
  type: AutoIncrementFieldType = 'auto-increment'
  system = true

  static create(input: Omit<ICreateAutoIncrementFieldInput, 'type'>): AutoIncrementField {
    const fieldName = FieldName.create(input.name)
    return new AutoIncrementField({
      id: FieldId.fromNullableString(input.id),
      name: fieldName,
      valueConstrains: FieldValueConstraints.create({ required: input.required }),
    })
  }

  static unsafeCreate(input: ICreateAutoIncrementFieldInput): AutoIncrementField {
    return new AutoIncrementField({
      id: FieldId.fromNullableString(input.id),
      name: FieldName.unsafaCreate(input.name),
      valueConstrains: FieldValueConstraints.unsafeCreate({ required: input.required }),
    })
  }

  createValue(value: ICreateAutoIncrementFieldValue): AutoIncrementFieldValue {
    return new AutoIncrementFieldValue(value)
  }

  createFilter(operator: IAutoIncrementFilterOperator, value: number | null): IAutoIncrementFilter {
    return { operator, value, path: this.id.value, type: 'auto-increment' }
  }

  accept(visitor: IFieldVisitor): void {
    visitor.autoIncrement(this)
  }
}
