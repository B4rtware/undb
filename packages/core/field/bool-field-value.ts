import { ValueObject } from '@egodb/domain'
import type { IBoolFieldValue } from './bool-field.type'

export class BoolFieldValue extends ValueObject<IBoolFieldValue> {
  constructor(value: IBoolFieldValue) {
    super({ value })
  }

  static T(): BoolFieldValue {
    return new this(true)
  }

  static F(): BoolFieldValue {
    return new this(false)
  }
}
