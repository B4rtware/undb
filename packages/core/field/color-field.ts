import type { IColorFilter, IColorFilterOperator } from '../filter/color.filter.js'
import { ColorFieldValue } from './color-field-value.js'
import type { ColorFieldType, ICreateColorFieldInput, ICreateColorFieldValue } from './color-field.type.js'
import { BaseField } from './field.base.js'
import type { IColorField } from './field.type.js'
import type { IFieldVisitor } from './field.visitor.js'
import { FieldId, FieldName, FieldValueConstraints } from './value-objects/index.js'

export class ColorField extends BaseField<IColorField> {
  type: ColorFieldType = 'color'

  static create(input: Omit<ICreateColorFieldInput, 'type'>): ColorField {
    const fieldName = FieldName.create(input.name)

    return new ColorField({
      id: FieldId.fromNullableString(input.id),
      name: fieldName,
      valueConstrains: FieldValueConstraints.create({ required: input.required }),
    })
  }

  static unsafeCreate(input: ICreateColorFieldInput): ColorField {
    return new ColorField({
      id: FieldId.fromNullableString(input.id),
      name: FieldName.unsafaCreate(input.name),
      valueConstrains: FieldValueConstraints.unsafeCreate({ required: input.required }),
    })
  }

  createValue(value: ICreateColorFieldValue): ColorFieldValue {
    return new ColorFieldValue(value)
  }

  createFilter(operator: IColorFilterOperator, value: string | null): IColorFilter {
    return { operator, value, path: this.id.value, type: 'color' }
  }

  accept(visitor: IFieldVisitor): void {
    visitor.color(this)
  }
}
