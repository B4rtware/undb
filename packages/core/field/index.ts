import { createNumberFieldSchema } from './number-field.type.js'
import { createStringFieldSchema } from './string-field.type.js'

export * from './auto-increment-field-value.js'
export * from './auto-increment-field.js'
export * from './bool-field-value.js'
export * from './bool-field.js'
export * from './color-field-value.js'
export * from './color-field.js'
export * from './color-field.type.js'
export * from './created-at-field-value.js'
export * from './created-at-field.js'
export * from './date-field-value.js'
export * from './date-field.js'
export * from './date-field.type.js'
export * from './date-range-field-value.js'
export * from './date-range-field.js'
export * from './date-range-field.type.js'
export * from './email-field-value.js'
export * from './email-field.js'
export * from './email-field.type.js'
export * from './field-value.visitor.js'
export * from './field.base.js'
export * from './field.constants.js'
export * from './field.factory'
export * from './field.type.js'
export * from './field.visitor.js'
export * from './id-field-value.js'
export * from './id-field.js'
export * from './number-field-value.js'
export * from './number-field.js'
export * from './parent-field-value.js'
export * from './parent-field.js'
export * from './parent-field.type.js'
export * from './rating-field-value.js'
export * from './rating-field.js'
export * from './rating-field.type.js'
export * from './reference-field-value.js'
export * from './reference-field.js'
export * from './select-field-value.js'
export * from './select-field.js'
export * from './select-field.type.js'
export * from './specifications/index.js'
export * from './string-field-value.js'
export * from './string-field.js'
export * from './tree-field-value.js'
export * from './tree-field.js'
export * from './tree-field.type.js'
export * from './updated-at-field-value.js'
export * from './updated-at-field.js'
export * from './value-objects/index.js'
export { createNumberFieldSchema, createStringFieldSchema }
