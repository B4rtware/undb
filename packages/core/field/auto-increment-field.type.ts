import * as z from 'zod'
import { AutoIncrementField } from './auto-increment-field'
import { baseFieldQuerySchema, createBaseFieldsSchema } from './field.base'
import { FIELD_TYPE_KEY } from './field.constant'

export const autoIncrementTypeSchema = z.literal('auto-increment')
export type AutoIncrementFieldType = z.infer<typeof autoIncrementTypeSchema>
const autoIncrementTypeObjectSchema = z.object({ [FIELD_TYPE_KEY]: autoIncrementTypeSchema })

export const createAutoIncrementFieldSchema = createBaseFieldsSchema.merge(autoIncrementTypeObjectSchema)
export type ICreateAutoIncrementFieldInput = z.infer<typeof createAutoIncrementFieldSchema>

export const autoIncrementFieldQuerySchema = baseFieldQuerySchema.merge(autoIncrementTypeObjectSchema)

export const autoIncrementFieldValue = z.number().int().positive()
export type IAutoIncrementFieldValue = z.infer<typeof autoIncrementFieldValue>

export const createAutoIncrementFieldValue = autoIncrementFieldValue
export type ICreateAutoIncrementFieldValue = z.infer<typeof createAutoIncrementFieldValue>

export const autoIncrementQueryValue = autoIncrementFieldValue
export type IAutoIncrementQueryValue = z.infer<typeof autoIncrementQueryValue>

export const createAutoIncrementFieldValue_internal = z
  .object({ value: createAutoIncrementFieldValue })
  .merge(autoIncrementTypeObjectSchema)
  .merge(z.object({ field: z.instanceof(AutoIncrementField) }))
export type ICreateAutoIncrementFieldValue_internal = z.infer<typeof createAutoIncrementFieldValue_internal>
