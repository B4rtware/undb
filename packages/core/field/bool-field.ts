import type { IBoolFilter } from '../filter/bool.filter'
import type { IBoolFilterOperator } from '../filter/operators'
import { BoolFieldValue } from './bool-field-value'
import type { BoolFieldType, ICreateBoolFieldInput, ICreateBoolFieldValue } from './bool-field.type'
import { BaseField } from './field.base'
import type { IBoolField } from './field.type'
import { FieldId, FieldKey, FieldName, FieldValueConstraints } from './value-objects'

export class BoolField extends BaseField<IBoolField> {
  type: BoolFieldType = 'bool'

  static create(input: ICreateBoolFieldInput): BoolField {
    const fieldName = FieldName.create(input.name)
    return new BoolField({
      id: FieldId.fromNullableString(input.id),
      key: input.key ? FieldKey.from(input.key) : FieldKey.fromName(fieldName),
      name: fieldName,
      valueConstrains: FieldValueConstraints.create({ required: input.required }),
    })
  }

  static unsafeCreate(input: ICreateBoolFieldInput): BoolField {
    return new BoolField({
      id: FieldId.fromNullableString(input.id),
      key: FieldKey.from(input.key),
      name: FieldName.unsafaCreate(input.name),
      valueConstrains: FieldValueConstraints.unsafeCreate({ required: input.required }),
    })
  }

  createValue(value: ICreateBoolFieldValue): BoolFieldValue {
    return new BoolFieldValue(value)
  }

  createFilter(operator: IBoolFilterOperator, value: boolean | null): IBoolFilter {
    return { operator, value, path: this.id.value, type: 'bool' }
  }
}
