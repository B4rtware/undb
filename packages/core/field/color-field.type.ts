import * as z from 'zod'
import { ColorField } from './color-field'
import { baseFieldQuerySchema, createBaseFieldsSchema } from './field.base'
import { FIELD_TYPE_KEY } from './field.constant'

export const colorTypeSchema = z.literal('color')
export type ColorFieldType = z.infer<typeof colorTypeSchema>
const colorTypeObjectSchema = z.object({ [FIELD_TYPE_KEY]: colorTypeSchema })

export const createColorFieldSchema = createBaseFieldsSchema.merge(colorTypeObjectSchema)
export type ICreateColorFieldInput = z.infer<typeof createColorFieldSchema>

export const colorFieldQuerySchema = baseFieldQuerySchema.merge(colorTypeObjectSchema)

export const colorFieldValue = z.string().min(4).max(9).regex(/^#/).nullable()
export type IColorFieldValue = z.infer<typeof colorFieldValue>

export const createColorFieldValue = colorFieldValue
export type ICreateColorFieldValue = z.infer<typeof createColorFieldValue>

export const colorFieldQueryValue = colorFieldValue
export type IColorFieldQueryValue = z.infer<typeof colorFieldQueryValue>

export const createColorFieldValue_internal = z
  .object({ value: createColorFieldValue })
  .merge(colorTypeObjectSchema)
  .merge(z.object({ field: z.instanceof(ColorField) }))
export type ICreateColorFieldValue_internal = z.infer<typeof createColorFieldValue_internal>
