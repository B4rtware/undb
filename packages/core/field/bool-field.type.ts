import * as z from 'zod'
import { BoolField } from './bool-field'
import { baseFieldQuerySchema, createBaseFieldsSchema } from './field.base'
import { FIELD_TYPE_KEY } from './field.constant'

export const boolTypeSchema = z.literal('bool')
export type BoolFieldType = z.infer<typeof boolTypeSchema>
const boolTypeObjectSchema = z.object({ [FIELD_TYPE_KEY]: boolTypeSchema })

export const createBoolFieldSchema = createBaseFieldsSchema.merge(boolTypeObjectSchema)
export type ICreateBoolFieldInput = z.infer<typeof createBoolFieldSchema>

export const boolFieldQuerySchema = baseFieldQuerySchema.merge(boolTypeObjectSchema)

export const boolFieldValue = z.boolean()
export type IBoolFieldValue = z.infer<typeof boolFieldValue>

export const createBoolFieldValue = boolFieldValue
export type ICreateBoolFieldValue = z.infer<typeof createBoolFieldValue>

export const createBoolFieldValue_internal = z
  .object({ value: createBoolFieldValue })
  .merge(boolTypeObjectSchema)
  .merge(z.object({ field: z.instanceof(BoolField) }))
export type ICreateBoolFieldValue_internal = z.infer<typeof createBoolFieldValue_internal>
