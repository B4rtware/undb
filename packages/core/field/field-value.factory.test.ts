import { FieldValueFactory } from './field-value.factory'
import { IFieldValue } from './field.type'

test.each<IFieldValue>(['hello', 1, null])('field value test', (value) => {
  const fieldValue = FieldValueFactory.from(value)
  expect(fieldValue).toMatchSnapshot()
})
