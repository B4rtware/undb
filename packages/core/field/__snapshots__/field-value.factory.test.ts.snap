// Vitest Snapshot v1

exports[`field value test 1`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): TextFieldValue {
    "props": {
      "value": "hello",
    },
  },
}
`;

exports[`field value test 2`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): NumberFieldValue {
    "props": {
      "value": 1,
    },
  },
}
`;

exports[`field value test 3`] = `
OptionType {
  Symbol(T): false,
  Symbol(Val): undefined,
}
`;